### ISSUE TITLE
Add SGDR, SGDW, AdamW,AdamWR

### ISSUE THREAD

**Issue by onlytailei** [pytorch/pytorch#3790]

**Comment by onlytailei** [comment]
https://github.com/pytorch/pytorch/pull/3740

**Comment by Kaixhin** [comment]
I think this (https://github.com/pytorch/pytorch/pull/3311) is the equivalent to TF's method, and if people want restarts then those can be build on top of cosine annealing?

**Comment by onlytailei** [comment]
@Kaixhin  
I didn't notice #3311  before. I will see if SGDW and AdamW can be built on that.

**Comment by anandsaha** [comment]
@onlytailei @jingweiz I have implemented this too and got [good results](https://github.com/anandsaha/fastai.part1.v2/blob/master/courses/dl1/cifar10.ipynb). 

I implemented it outside PyTorch in fastai library, and was to put it in PyTorch. 

How far are you guys with it?

**Comment by Kaixhin** [comment]
@anandsaha if you could have a look at my PR (linked above) and check that it is fine then that would be useful. In particular I've also written tests for it, so once the PyTorch devs are free to check it then it can hopefully be merged (but there's lots of higher priority work going on with the internals of PyTorch atm).

Since we have a learning rate scheduler module, it would be good to be modular and use that for (warm) restarts. I haven't checked in detail but I imagine that a `torch.optim.lr_scheduler.WarmRestartLR` could take another scheduler and a restart schedule as arguments, and then it could be applied to any optimiser, and potentially and learning rate decay scheme (the docs can have the info and example that the usage of cosine annealing is what would be needed to replicate that paper).

**Comment by anandsaha** [comment]
@Kaixhin Sure! I will take a look.

**Comment by lucasb-eyer** [comment]
See also #4429 for the weight-decay part only.

**Comment by jpeg729** [comment]
If we choose to not fix the current implementations of Adam, then at the very least we should add a warning saying that if you want weight_decay as in the original proposal, then you need to use AdamW.

What about adding a `true_weight_decay` argument to Adam?

**Comment by weiyangfb** [comment]
also see AdamWR: https://github.com/pytorch/pytorch/pull/3740

**Comment by ogrisel** [comment]
> What about adding a true_weight_decay argument to Adam?

The [ICLR 2019 update of the AdamW paper](https://arxiv.org/abs/1711.05101) was renamed to "Decoupled Weight Decay Regularization". If this is to be included as an option to the default Adam optimizer, I think it should be named `decoupled_weight_decay` instead of `true_weight_decay`.

**Comment by soumith** [comment]
@vincentqb please review this at your convenience as well. i think it overlaps with the other PRs.
